### ISSUE TITLE
Cursor(Stream) constructor has "short read", doesn't check return value from Stream.Read()

### ISSUE THREAD

**Issue by rickbrew** [dotnet/winforms#6798]
This is in .NET 6.0.2 but I don't believe it's new, it may be a very old issue.

Paint.NET 4.3.8 (latest public release) started crashing on the new Windows 11 Insider "Dev" build, 10.0.22567.0. It seems to be because of some `Cursor`s that aren't initializing correctly. My cursor files haven't changed in quite some time, nor has the code for handling them. Now, I do believe this is a new bug in Win11, but there also seems to be a bug in WinForms.

Filip Navara @filipnavara on Twitter pointed out ( https://twitter.com/filipnavara/status/1499792311965913093?s=20&t=TUXrCwLBMiomnfPGQecZEw ) a "short read" in the `Cursor(Stream)` constructor: https://github.com/dotnet/winforms/blob/04938f30a978874cebc88a399efbd0e89e1b4cb8/src/System.Windows.Forms/src/System/Windows/Forms/Cursor.cs#L87

So the data buffer is incompletely read into (because `Stream.Read()` only promises to read more than 0 but less than or equal to `length` bytes), initializing the handle fails, but the code keeps marching on

Here's the exception and callstack that's filling up my inbox, all from Win11 Insider Dev build 10.0.22567.0: 
```
Exception details:
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Disposed 'Cursor' can no longer be used.'.
   at System.Windows.Forms.Cursor.get_Handle()
   at Interop.User32.SetCursor(IHandle hCursor)
   at System.Windows.Forms.Control.WmSetCursor(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
```

cc @bitcrazed @filipnavara @RussKie

**Comment by rickbrew** [comment]
Looks like this bug was reported two years ago and hasn't been fixed yet :-\ https://github.com/dotnet/winforms/issues/3492

**Comment by rickbrew** [comment]
To repro a crash:

1. Install Win11 Insider Dev build 10.0.22567.0 (currently the latest one)
2. Install Paint.NET 4.3.8 (or use portable build from here: https://github.com/paintdotnet/release/releases )
3. Launch Paint.NET
4. Click on Eraser tool (can press 'E')
5. Click in the image area

**Comment by rickbrew** [comment]
I'm currently testing a workaround (copy to `MemoryStream` first, then hand off to `Cursor(Stream)`), and I should be able to release an app update (v4.3.9) once .NET 6.0.3 is published (presumably on Tuesday?)

**Comment by roland5572** [comment]
@rickbrew This is a known issue, you can check this link for more useful information. This problem has existed since 2016 year and is still unsolved.

https://github.com/dotnet/runtime/issues/16598

**Comment by trontronicent** [comment]
I think this just have gotten worse, with the latest insider build. Even a standard property grid with a simple 'SelectedObject' set - after moving the mouse around, the exact same Cursor disposed / ObjectDisposedException is risen , leading to a ThreadException - as this is happening within the OnMouseMove event of PropertyGridInternal.

And to make it even worse: Click the Cursors property of a property grid within the latest VS2022 Preview, scroll down to the empty Cursor Elements and see Visual Studio crash. 

Then open EventViewer of Windows  and see it crash with the same Exception

EventViewer
Anwendung: mmc.exe
Frameworkversion: v4.0.30319
Beschreibung: Der Prozess wurde aufgrund einer unbehandelten Ausnahme beendet.
Ausnahmeinformationen: System.ObjectDisposedException
   bei System.Windows.Forms.Cursor.get_Handle()
   bei System.Windows.Forms.RichTextBox.WndProc(System.Windows.Forms.Message ByRef)
   bei System.Windows.Forms.NativeWindow.Callback(IntPtr, Int32, IntPtr, IntPtr)

Ausnahmeinformationen: System.Reflection.TargetInvocationException
   bei Microsoft.ManagementConsole.Internal.SnapInMessagePumpProxy.OnThreadException(System.Object, System.Threading.ThreadExceptionEventArgs)
   bei System.Windows.Forms.Application+ThreadContext.OnThreadException(System.Exception)
   bei System.Windows.Forms.Control.WndProcException(System.Exception)
   bei System.Windows.Forms.NativeWindow.Callback(IntPtr, Int32, IntPtr, IntPtr)

VS
From Related VS Crash:
Anwendung: devenv.exe
Beschreibung: Der Prozess wurde aufgrund einer unbehandelten Ausnahme beendet.
Ausnahmeinformationen: System.ObjectDisposedException
   bei System.Windows.Forms.Cursor.get_Handle()
   bei System.Windows.Forms.Cursor.DrawImageCore(System.Drawing.Graphics, System.Drawing.Rectangle, System.Drawing.Rectangle, Boolean)
   bei System.Windows.Forms.Cursor.DrawStretched(System.Drawing.Graphics, System.Drawing.Rectangle)
   bei Microsoft.WinForms.DesignTools.Client.Editors.CursorEditor+CursorUI.OnDrawItem(System.Windows.Forms.DrawItemEventArgs)
   bei System.Windows.Forms.ListBox.WmReflectDrawItem(System.Windows.Forms.Message ByRef)
   bei System.Windows.Forms.ListBox.WndProc(System.Windows.Forms.Message ByRef)
   bei System.Windows.Forms.NativeWindow.DebuggableCallback(IntPtr, Int32, IntPtr, IntPtr)
   
   
   I think, as it now affects also a stock control + spreads even to the IDE & the OS itself, it's about time....

**Comment by marianodrzazga** [comment]
Today Win11 updated and I started having this issue. Is there any workaround for this situation? In my case, I get the exception every time I hover the split container splitter bar (either un runtime, or with the VS designer, which forces VS to crash)

```   at System.Windows.Forms.Cursor.get_Handle()
   at Interop.User32.SetCursor(IHandle hCursor)
   at System.Windows.Forms.SplitContainer.WmSetCursor(Message& m)
   at System.Windows.Forms.SplitContainer.WndProc(Message& msg)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.SendMessageW(IntPtr hWnd, WM Msg, IntPtr wParam, IntPtr lParam)
   at Interop.User32.SendMessageW(IHandle hWnd, WM Msg, IntPtr wParam, IntPtr lParam)
   at System.Windows.Forms.SplitContainer.set_OverrideCursor(Cursor value)
   at System.Windows.Forms.SplitContainer.OnMouseMove(MouseEventArgs e)
   at System.Windows.Forms.Control.WmMouseMove(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ScrollableControl.WndProc(Message& m)
   at System.Windows.Forms.SplitContainer.WndProc(Message& msg)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, WM msg, IntPtr wparam, IntPtr lparam)
   at Interop.User32.DispatchMessageW(MSG& msg)
   at System.Windows.Forms.Application.ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop(UIntPtr dwComponentID, msoloop uReason, Void* pvLoopData)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(msoloop reason, ApplicationContext context)
   at System.Windows.Forms.Application.ThreadContext.RunMessageLoop(msoloop reason, ApplicationContext context)
   at System.Windows.Forms.Application.Run(Form mainForm)
   at Ionik.Windows.Client.Program.Main() in C:\code\md\ionik\Ionik.Client\Client\Program.cs:line 47

**Comment by bitcrazed** [comment]
Hi all. I can confirm that this is a bug in Win11 Insiders Dev release 10.0.22567.0, not in Paint.NET.

We have a fix and are working to get it released into a future Win11 Insiders Dev build ASAP.

Apologies for the inconvenience this issue has caused.

**Comment by AngeloCresta** [comment]
In version 10.0.22572.1 the bug is still present

**Comment by dreddy-work** [comment]
@bitcrazed can you please share expected windows version with fix here?

**Comment by bitcrazed** [comment]
> @bitcrazed can you please share expected windows version with fix here?

Will do as soon as I can confirm which build the fix will arrive in. Stay tuned!

**Comment by weltkante** [comment]
> Looks like this bug was reported two years ago and hasn't been fixed yet 

just for the record and people referencing #3492, which was a bug about custom streams and never affected cursor loads from files

File streams were always "implicitly guaranteed" to read as much as possible on Windows, too much code relied on that behavior historically to change it, I'd be surprised if .NET team came to a different conclusion now and decided to break people.

Reading the last few comments sounds like it has been confirmed to be an issue external to WinForms?

**Comment by rickbrew** [comment]
> File streams were always "implicitly guaranteed" to read as much as possible on Windows, too much code relied on that behavior historically to change it, I'd be surprised if .NET team came to a different conclusion now and decided to break people.

Yes but .NET 6 introduced a breaking change to `GZipStream` (and others) such that they do not guarantee full reads. This bug here is a landmine and should be pretty easy to fix.

> Reading the last few comments sounds like it has been confirmed to be an issue external to WinForms?
Yes, the particular crash that's being reported (here and elsewhere) is on current Win11 Insider Dev builds. I think the fix is still making its way into the proper branch.

**Comment by RussKie** [comment]
> This bug here is a landmine and should be pretty easy to fix.

Would you mind sending a PR with a fix?

**Comment by JosefPapug** [comment]
Same problem with e.g. Axure application. Still bad in 22572.

**Comment by AngeloCresta** [comment]
The external issue has been fixed in Version 10.0.22579.1 (Dev Channel), and, from release notes:
_Fixed an issue which was causing the mouse to be invisible in fields of certain apps, as well as some unexpected cursor related app crashes in recent flights._
[https://blogs.windows.com/windows-insider/2022/03/18/announcing-windows-11-insider-preview-build-22579/](url)

**Comment by RussKie** [comment]
Closing as the underlying issues is fixed.
